build-ui: Extract protobuf clients into js packages

dhartunian opened the pull request:
This PR is the first step to extracting the Admin UI from the CockroachDB repository. This is to enable more rapid iteration on the UI and to enable embedding it into other applications.

The purpose of these changes is to establish a clear boundary between the generated protobuf client code and the Admin UI app.

The existing configuration would place the generated code in the `/src` directory of the app itself and compile it just like the rest of the hand-written code. Once the Admin UI is extracted, we cannot take the `.proto` files with us since the DB is the proper owner. Hence we need a library wrapper around the generated clients to easily import them outside of this repository.

We've chosen to wrap the JS clients in simple npm repositories for now. Once the Admin UI is extracted, we will be publishing these libraries to NPM. For now, they are symlinked dependencies in the `/pkg/ui` codebase.

This change also eliminates the `proto` javascript build process since the protobuf clients are wrapped into the `vendor` build like all other dependencies.

This closes #43938

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44427)
<!-- Reviewable:end -->

dhartunian committed:
build-ui: Add ui-client packages to ui build

- Adding 2 new packages as linked imports (node_modules just gets
  symlinked)
- Adjusted webpack vendor build to bundle the new imports properly
  since they're not independently run through babel
- Adjusted webpack vendor build to omit `ccl` package on ccl builds

Release note: none

dhartunian committed:
build-ui: migrate to using new protobuf libraries in code

- remove references to the protos DLL
- remove webpack references to protos DLL
- adjust Makefile to put proto files in new locations
  - removed the ccl/oss distinction for dll outputs since
    we just have a single vendor file that's output for both

Release note: none

Koorosh commented on Makefile:
```diff
@@ -1308,8 +1308,8 @@ ui-lint: pkg/ui/yarn.installed $(UI_PROTOS_OSS) $(UI_PROTOS_CCL)
-UI_CCL_DLLS := pkg/ui/dist/protos.ccl.dll.js pkg/ui/dist/vendor.oss.dll.js
-UI_CCL_MANIFESTS := pkg/ui/protos.ccl.manifest.json pkg/ui/vendor.oss.manifest.json
+UI_CCL_DLLS := pkg/ui/dist/vendor.oss.dll.js
+UI_CCL_MANIFESTS := pkg/ui/vendor.oss.manifest.json
```
Is it a naming mismatch for variable and library? UI_**CCL**_MANIFESTS --> pkg/ui/vendor.**oss**.manifest.json
CCL in variable name refers to OSS file name.

dhartunian replied:
Thanks @Koorosh nice catch, that does look like an naming mismatch!

dhartunian replied:
After taking a closer look, this is, in fact, correct. The vendor file is *always* marked as `oss` since everything in it is open source. I've updated the Makefile to rename these variables and make that clearer.

Koorosh commented on pkg/ui-client/oss/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+    "noImplicitAny": true,
+    "removeComments": true,
+    "preserveConstEnums": true,
+    "outFile": "./dist/index.d.ts"
```
As I know, `outFile` is used to indicate output files not file definitions. Currently it indicates on '.d.ts' file. Is it intentionally?

dhartunian replied:
not intentional, I might take this out completely now that you've flagged it. We don't really do any TS compilation in these tiny repos.

dhartunian replied:
config has been removed since no TS compilation is happening in these projects

celiala reviewed (commented):
Woo! Super excited for this project!

On a high-level read, everything here makes sense to me.

David and I sync'ed offline about this already: trying to `make build` this fails because `pkg/ui-client/ccl/src/` is empty (running `mkdir pkg/ui-client/ccl/src/` should work). I'll try this again in a bit.

dhartunian committed:
build-ui: Add new UI packages for protobuf libraries

In order to lift out the Admin UI codebase from CRDB, we need to establish
clearer boundaries around CRDB and UI code. The protobuf Javascript and
Typescript files define this boundary and must remain in this repo.

The protobuf client will become a JS library just like any other Admin UI
dependency.

For now these are just skeleton repos. Future commits will incorporate their
output into the UI build.

Release note: None

dhartunian commented:
@celiala pushed changes that add little READMEs into the `src` directories to ensure they're created and kept to hold the output of the protobuf-js process.
